src/hotspot/share/opto/vmaskloop.hpp

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) 2023, Arm Limited. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#ifndef SHARE_OPTO_VMASKLOOP_HPP
+#define SHARE_OPTO_VMASKLOOP_HPP
+
+#include "opto/loopnode.hpp"
+#include "opto/superword.hpp"
+#include "opto/vectornode.hpp"
+
+// ----------------------------- VectorMaskedLoop -----------------------------
+class VectorMaskedLoop : public ResourceObj {
+ private:
+  // Useful handles
+  PhaseIdealLoop*            _phase;
+  PhaseIterGVN*              _igvn;
+  Arena*                     _arena;
+
+  // Loop information
+  IdealLoopTree*             _lpt;          // Idealloop tree
+  CountedLoopNode*           _cl;           // CountedLoop node
+  CountedLoopEndNode*        _cle;          // CountedLoopEnd node
+  PhiNode*                   _iv;           // Loop induction variable PhiNode
+
+  // Data structures for loop analysis
+  Unique_Node_List           _core_set;     // Loop core nodes set for fast membership check
```

Comment:
If this is really only for membership test, and you never need the list of nodes, you could just use the `VectorSet`. Uses less memory.